### PR TITLE
KAFKA-15181: fix(storage): wait for consumer to be synced after assigning partitions

### PR DESCRIPTION
**Pull request by jeqo**
This fixes a race condition where cache is not ready but TBRLMM is marked as initialized anyway, causing replica to fail as remote log metadata is not found.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mdedetrich** [comment]
Pinging @satishd and @junrao since its tiered storage related.

**Review by mdedetrich — APPROVED** [review]
Implementation looks good with me, also appreciate the minor code improvements. Only question is whether a test is required for this, and if so is there some prior art for testing `waitTillConsumptionCatchesUp`.

We are also going to test this implementation internally to verify there are no issues.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 1]
```diff
@@ -78,45 +88,42 @@ public void startConsumerThread() {
     /**
      * Waits if necessary for the consumption to reach the offset of the given {@code recordMetadata}.
      *
-     * @param recordMetadata record metadata to be checked for consumption.
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:81]: Please remember to update the javadoc.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 2]
```diff
@@ -78,45 +88,42 @@ public void startConsumerThread() {
...
+        waitTillConsumptionCatchesUp(partition, offset, rlmmConfig.consumeWaitMs());
     }
 
     /**
      * Waits if necessary for the consumption to reach the offset of the given {@code recordMetadata}.
      *
-     * @param recordMetadata record metadata to be checked for consumption.
-     * @param timeoutMs      wait timeout in milli seconds
+     * @param timeoutMs wait timeout in milliseconds
      * @throws TimeoutException if this method execution did not complete with in the given {@code timeoutMs}.
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: javadoc needs to be updated.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 3]
```diff
@@ -78,45 +88,42 @@ public void startConsumerThread() {
...
-                                             long timeoutMs) throws TimeoutException {
-        final int partition = recordMetadata.partition();
+    public void waitTillConsumptionCatchesUp(int partition, long offset, long timeoutMs) throws TimeoutException {
         final long consumeCheckIntervalMs = Math.min(CONSUME_RECHECK_INTERVAL_MS, timeoutMs);
 
         // If the current assignment does not have the subscription for this partition then return immediately.
         if (!consumerTask.isPartitionAssigned(partition)) {
-            throw new KafkaException("This consumer is not subscribed to the target partition " + partition + " on which message is produced.");
+            throw new KafkaException("This consumer is not subscribed to the target partition " + partition + " assigned. " +
+                    "Partitions currently assigned: " + consumerTask.partitionsAssigned());
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: Thanks for adding more context for the error message. But I don't understand why this is added:
`This consumer is not subscribed to the target partition " + partition + [" assigned. "]` <-- why?

File: storage/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -19,4 +19,4 @@ log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
 log4j.appender.stdout.layout.ConversionPattern=[%d] %p %m (%c:%L)%n
 
 log4j.logger.org.apache.kafka.server.log.remote.storage=INFO
-log4j.logger.org.apache.kafka.server.log.remote.metadata.storage=INFO
+log4j.logger.org.apache.kafka.server.log.remote.metadata.storage=DEBUG
```
Comment (showuon) [inline storage/src/test/resources/log4j.properties]: I don't we should default enable `DEBUG` log level. Should we revert this change?

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 5]
```diff
@@ -139,6 +147,33 @@ public void close() throws IOException {
...
     public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
         consumerTask.addAssignmentsForPartitions(partitions);
+        waitTillMetaPartitionIsInSync();
+    }
+
+    /**
+     * Whenever there is a change on the meta partitions assgined to the TBRLMM,
+     * wait for consumer (therefore the cache as well) to be in-sync before continuing
+     */
+    private void waitTillMetaPartitionIsInSync() {
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: Could we change the term `inSync` to `catchUp` as above?

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 6]
```diff
@@ -139,6 +147,33 @@ public void close() throws IOException {
...
+
+                for (TopicPartition tp : topicPartitions) {
+                    waitTillConsumptionCatchesUp(
+                            tp.partition(),
+                            latestOffsets.get(tp).offset() - 1 // as latest offset is the next latest written record
+                    );
+                }
+            }
+        } catch (ExecutionException | InterruptedException | TimeoutException e) {
+            log.error("Error encountered while consuming from meta partitions", e);
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: I think we should output the partition name, and remove the ambiguous "meta partition", ex:
`("Error encountered while consuming from remote log metadata partitions: {}", consumerTask.partitionsAssigned(), e)`

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java [hunk 7]
```diff
@@ -363,7 +363,7 @@ private void initializeResources() {
         long startTimeMs = time.milliseconds();
         AdminClient adminClient = null;
         try {
-            adminClient = AdminClient.create(rlmmConfig.producerProperties());
+            adminClient = AdminClient.create(rlmmConfig.commonProperties());
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java]: Thanks for the improvement. Could we use try-with-resource here?

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 8]
```diff
@@ -139,6 +147,33 @@ public void close() throws IOException {
 
     public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
         consumerTask.addAssignmentsForPartitions(partitions);
+        waitTillMetaPartitionIsInSync();
+    }
+
+    /**
+     * Whenever there is a change on the meta partitions assgined to the TBRLMM,
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: Could we make this clear? ex:
`Whenever there is a change on the remote log metadata topic partitions assgined to the topic based remote log metadata manager,`

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java (hunk 6 above)
Comment (showuon) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: Should we throw exception here?

**Review by showuon — COMMENTED** [review]
@jeqo , nice find and fix! Left some comments.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java (hunk 3 above)
Comment (jeqo) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java]: Sure, let me rephrase that to 
```
            throw new KafkaException("This consumer is not assigned to the target partition " + partition + ". " +
                    "Partitions currently assigned: " + consumerTask.metadataPartitionsAssigned());

```
as partitions are not technically subscribed, but manually assigned.

File: storage/src/test/resources/log4j.properties (hunk 4 above)
Comment (jeqo) [reply storage/src/test/resources/log4j.properties]: Enabled it because it was hard to troubleshoot test issues without it; also only enabled on tests. Though I'm ok reverting it.

**Review by showuon — COMMENTED** [review]
One more comment, I don't see tests for the new added method. Could you add some tests for them?

**Comment by jeqo** [comment]
@showuon thanks for your feedback! I have applied most suggestions -- let me know if I'm missing anything.

About the test, I followed @mdedetrich advise on piggybacking on existing testing of `waitTillConsumptionCatchesUp` and actually removing the waiting from `TopicBasedRemoteLogMetadataManagerTest` to prove the new behavior when assigning partitions. 

I tried to add some scenarios to `TopicBasedRemoteLogMetadataManagerRestartTest` but the current threads initialization haven't been able to reproduce. Maybe something to improve in a follow up PR, I think this one may be related: https://github.com/apache/kafka/pull/13689

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 9]
```diff
@@ -76,47 +94,50 @@ public void startConsumerThread() {
...
+     * @param partition partition of metadata topic to be checked for consumption
+     * @param offset record offset of metadata topic {@code partition} to be checked for consumption
+     * @param timeoutMs wait timeout in milliseconds
      * @throws TimeoutException if this method execution did not complete with in the given {@code timeoutMs}.
      */
-    public void waitTillConsumptionCatchesUp(RecordMetadata recordMetadata,
-                                             long timeoutMs) throws TimeoutException {
-        final int partition = recordMetadata.partition();
+    public void waitTillConsumptionCatchesUp(int partition, long offset, long timeoutMs) throws TimeoutException {
         final long consumeCheckIntervalMs = Math.min(CONSUME_RECHECK_INTERVAL_MS, timeoutMs);
```
Comment (showuon) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:121]: I think we should log something as INFO when waiting the consumer catching up, ex:
`log.info("waiting until consumer is caught up with the target partition {}", partition);`
so that we can know why it's pending without enabling debug log. WDYT?

**Review by showuon — COMMENTED** [review]
LGTM, just a minor comment.

> About the test, I followed @mdedetrich advise on piggybacking on existing testing of waitTillConsumptionCatchesUp and actually removing the waiting from TopicBasedRemoteLogMetadataManagerTest to prove the new behavior when assigning partitions.

Fair enough!

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the fix!

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java [hunk 10]
```diff
@@ -209,7 +214,7 @@ private void maybeSyncCommittedDataAndOffsets(boolean forceSync) {
                     if (offset != null) {
                         remotePartitionMetadataEventHandler.syncLogMetadataSnapshot(topicIdPartition, metadataPartition, offset);
                     } else {
-                        log.debug("Skipping syncup of the remote-log-metadata-file for partition:{} , with remote log metadata partition{}, and no offset",
+                        log.debug("Skipping sync-up of the remote-log-metadata-file for partition:{} , with remote log metadata partition{}, and no offset",
```
Comment (kamalcph) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:217]: partition{} -> partition {}

**Review by kamalcph — COMMENTED** [review]
@jeqo Thanks for the PR!

1. Can we split this patch into two?  a) Refactoring the existing code and b) Specific changes that address KAFKA-15181 ticket.
2. We are instantiating `AdminClient` (one more client) to do listOffset calls in ConsumerManager to ensure that once the user-partitions are assigned, all the messages are caught-up from the respective metadata-partition. 

Both Consumer and Admin clients, can do list-offset calls. Instead, of adding one more Admin client, can we move the initialisation logic to RemoteLogMetadataCache.java class by introducing a latch. All the calls to read/write from remote will hit the RemoteLogMetadataCache and the `ConsumerTask` can mark the partition as initialised once it reads all the messages.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java [hunk 11]
```diff
@@ -139,6 +163,35 @@ public void close() throws IOException {
 
     public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
         consumerTask.addAssignmentsForPartitions(partitions);
+        waitTillMetaPartitionConsumptionCatchesUp();
```
Comment (kamalcph) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: This is a blocking call. It won't return until the partition is ready/consumption-catches-up. In high throughput cluster, this can anywhere take order of seconds to complete. 

Moreover, when you reassign the partitions, the controller will send the LISR request for one partition at a time and the LISR request handling path shouldn't have any blocking call so that the controller can send the next LISR request on time. Otherwise, LISR request gets timed out.

Comment (showuon) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: Good point @kamalcph ! Do you have any suggestion here?

File: storage/src/test/resources/log4j.properties (hunk 4 above)
Comment (mdedetrich) [reply storage/src/test/resources/log4j.properties]: Agreed, we shouldn't default to `DEBUG` for production code otherwise it produces way too many logs.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java (hunk 9 above)
Comment (mdedetrich) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:121]: This makes sense especially since we are dealing with remote storage and hence we may get high lag times.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java (hunk 11 above)
Comment (mdedetrich) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: I am also unclear as to what we should do here, I mean we could return a `CompletableFuture<Void>` but this would involve changing the interface?

Comment (kamalcph) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: One approach we can take is to wait for metadata partition to catch up asynchronously. All the calls to read the remote log will hit the [RemotePartitionMetadataStore#getRemoteLogMetadataCache](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java#L147), there we can ensure that the partition is initialised via a latch. 

The [ConsumerTask](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java) can keep a marker of the latest offset of metadata-partition when user-partitions are assigned to them and can mark that user-partition as initialised in the cache once it read/re-read messages up-to that point. 

We can have a latch with timeout of 2 mins to avoid holding the caller thread for a long amount of time and throw a retry-able error.

Comment (showuon) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: Sounds great! I like the idea that we lazily wait for the consumer to catch up. One comment is why don't we return retriable error immediately, and ask the client to try it again, like when the leader is under election, the client will get `NOT_LEADER_OR_FOLLOWER` error immediately and retry, instead of waiting in the broker side until leader election completed. WDYT?

Comment (kamalcph) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: That also works! We have to look around for all the callers:

1. Replica fetcher thread will retry the request after the `replica.fetch.backoff.ms` once the partition is marked as error/failed.
2. Consumers may resend the FETCH request as soon as it fails without the request quota. 

We can also think about adding one more API in [RemoteLogMetadataManager](https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java) to ensure that the partition is initialised before copying the segments (or) when the replica thread is being initialised. 

```
boolean isInitialized(TopicIdPartition topicIdPartition) throws RemoteStorageException;
```

Comment (satishd) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:166]: The latching mechanism as mentioned by @kamalcph is what we are currently using internally. @abhijeetk88 is planning to pull those changes to trunk and raise PR this week to handle the scenario.

**Review by satishd — COMMENTED** [review]
Thanks @jeqo for the PR. Let us keep the changes separate as @kamalcph suggested in other [comment](https://github.com/apache/kafka/pull/14012#pullrequestreview-1533335715). We should avoid blocking the calls while registering a partition to RLMM, which is handled using latches in internal branch, left a [comment](https://github.com/apache/kafka/pull/14012#discussion_r1266471753) on the same. Let us have this PR with the cleanup and other minor changes.

**Comment by jeqo** [comment]
Thanks @satishd and @kamalcph, I've moved refactoring changes to https://github.com/apache/kafka/pull/14045, please take a look.

About the fix, looking forward to take a look to your PR. Will hold changes on this PR then, many thanks.

**Comment by satishd** [comment]
Thanks @jeqo for moving the changes to https://github.com/apache/kafka/pull/14045. As I already merged that PR, closing this PR.
